kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -1,173 +1,146 @@
     this.log = logPrefix.logger(AsyncThreadPool.class);
-
-    final ProcessingQueue processingQueue = new ProcessingQueue(logPrefix);
-    this.processingQueue = processingQueue;
-    startThreads(processingQueue);
+    // todo: set the thread names
+    executorService = Executors.newFixedThreadPool(size, r -> {
+      final var t = new Thread(r);
+      t.setDaemon(true);
+      t.setName(
+          streamThreadName + "-" + ASYNC_THREAD_NAME + "-" + threadNameIndex.getAndIncrement());
+      return t;
+    });
   }
 
-  private void startThreads(final ProcessingQueue processingQueue) {
-    for (int asyncThreadIndex = 0; asyncThreadIndex < threadPoolSize; ++asyncThreadIndex) {
-      final String name = generateAsyncThreadName(streamThreadName, asyncThreadIndex);
-      final AsyncThread thread = new AsyncThread(name, processingQueue);
-
-      thread.setDaemon(true);
-      threadPool.put(name, thread);
-    }
+  @VisibleForTesting
+  Map<AsyncEvent, Future<?>> getInFlight(final String processorName, final int partition) {
+    return inFlight.get(InFlightWorkKey.of(processorName, partition));
+  }
 
-    for (final AsyncThread thread : threadPool.values()) {
-      thread.start();
+  public void removeProcessor(final String processorName, final int partition) {
+    log.info("clean up records for {}:{}", processorName, partition);
+    final var key = InFlightWorkKey.of(processorName, partition);
+    final Map<AsyncEvent, Future<?>> inFlightForTask = inFlight.remove(key);
+    if (inFlightForTask != null) {
+      inFlightForTask.values().forEach(f -> f.cancel(true));
     }
-
-    log.info("Started up all {} async threads in this pool", threadPoolSize);
   }
 
-  public <KOut, VOut> void addProcessor(
-      final String asyncProcessorName,
-      final int partition,
-      final AsyncUserProcessorContext<KOut, VOut> userContext,
-      final FinalizingQueue finalizingQueue
+  public <KOut, VOut> void scheduleForProcessing(
+      final String processorName,
+      final int taskId,
+      final List<AsyncEvent> events,
+      final FinalizingQueue finalizingQueue,
+      final ProcessingContext originalContext,
+      final AsyncUserProcessorContext<KOut, VOut> asyncProcessorContext
   ) {
-    processingQueue.addPartition(partition);
-
-    for (final AsyncThread thread : threadPool.values()) {
-      final AsyncThreadProcessorContext<KOut, VOut> threadLocalContext =
-          new AsyncThreadProcessorContext<>(userContext.taskContext());
-
-      final AsyncNodeContainer processorContainer = new AsyncNodeContainer(
-          streamThreadName,
-          asyncProcessorName,
-          partition,
-          threadLocalContext,
-          finalizingQueue
-      );
-
-      thread.addProcessor(processorContainer, threadLocalContext, userContext);
+    // todo: can also check in-flight for failed tasks
+    for (final var e : events) {
+      final Future<?> future = executorService.submit(new AsyncEventTask<>(
```

Comment:
I guess I'm also a little uncomfortable with the idea of losing control over the ProcessingQueue. Do we really want to hand over all scheduling to the ExecutorService? Is there any way to customize the scheduling algorithm? I did actually start the POC with an executor service pool but replaced it with the custom threads for some reason, and I think this had something to do with it.

For example: going forward we might want to be smarter about prioritizing certain keys and/or partitions based on how relative weight in the set of blocked tasks. We may also be in a situation where Streams is attempting to close one of its tasks and commit only that one, in which case we should flush all the events for that partition first/with priority. What worries me most is that we just don't even know what a good scheduling algorithm will be and/or how we might want to apply heuristics to improve it under specific conditions or in extreme cases. So I guess I'd prefer to leave that door open.

idk. As always, maybe I'm just being paranoid here. Is there a way to use the ExecutorService without giving up control over the scheduling altogether? (Or is giving up control over scheduling more or less the point of using the ExecutorService?)

Doesn't mean we can't do it now, we should optimize for stability and production-readiness over theoreticals. Just trying to understand the stakes a bit bettter

Reply:
Yep, that all makes sense and was something I thought about a bit as well. In general I'm in favor of removing complexity and adding it if/when we actually need to solve those problems. That said, the native pool is pretty flexible and if we need to we can provide our own queue implementation to the thread pool, which I think lets us do all the stuff that you're describing: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html#ThreadPoolExecutor-int-int-long-java.util.concurrent.TimeUnit-java.util.concurrent.BlockingQueue-

Reply:
Fair enough. The only other thing I was worried about is the error handling aspect and how to deal with failures and user/processors exceptions, but I think you've pointed out in this PR how we can go about that. So I'm starting to get on board with the ExecutorService approach. Will give the PR another pass but I can agree with the general approach for now

Reply:
We have a couple of options for how to do that:
1. We can just have our task implementation catch exceptions thrown by `process` and stick them in the event when we append it to the finalizing queue. And then the finalizer can deal with them.
2. We can have the caller of `AsyncThreadPool` check the open futures for any failed ones and then deal with the errors then.

(1) seems the most straightforward to me.

Reply:
Agreed -- I think the more context we can keep scoped to a specific event, the better. Will help us provide useful diagnostic info as well